presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -250,9 +344,49 @@ public static PlanNode pushPredicateIntoTableScan(
             SqlParser parser,
             ExpressionDomainTranslator domainTranslator)
     {
+        if (metadata.isPushdownFilterSupported(session, node.getTable())) {
+            Map<NodeRef<Expression>, Type> predicateTypes = getExpressionTypes(
+                    session,
+                    metadata,
+                    parser,
+                    types,
+                    predicate,
+                    emptyList(),
+                    WarningCollector.NOOP,
+                    false);
+
+            BiMap<String, String> symbolToColumnNameMap = node.getAssignments().entrySet().stream()
+                    .collect(toImmutableBiMap(
+                            entry -> entry.getKey().getName(),
+                            entry -> getColumnName(session, metadata, node.getTable(), entry.getValue())));
+            RowExpression translatedPredicate = SqlToRowExpressionTranslator.translate(predicate, predicateTypes, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), session, false)
+                    .accept(new TranslateVariableNamesVisitor(), symbolToColumnNameMap);
+
+            PushdownFilterResult pushdownFilterResult = metadata.pushdownFilter(session, node.getTable(), translatedPredicate);
```

Comment:
I haven't gone into HiveConnector's code but it looks like it will replace the old `pickTableLayout`. There is a few things likely needs to be careful. One thing is that there is a few hacks in push down predicate on partition keys to do some evaluation on complicated predicate and not return huge tupleDomain. That's why constraint is passed in with a  `isCandidate`. 

Perhaps, that is something who knows HiveConnector can chime in.

Reply:
Another place might need to change is `MetadataQueryOptimizer`. Otherwise, we will has to keep both methods in `HiveConnector`.

Reply:
@hellium01 Yi, indeed, there are a number of places that invoke Metadata.getLayout and all of these will be need to be modified to use pushdownFilter as well. One option would be to fold filterPushdown into Metadata.getLayout. I plan on working through this in a separate PR, because I feel this PR is pretty big already. I would also like to set things up so that multiple people could work in parallel, e.g. once this PR lands, one person can work on the above while another one can work on subfield pruning and/or implementing filter evaluation in Hive connector.

Re: isCandidate - My understanding is that the new code path (via pushdownFilter) is missing out on some opportunities to prune partitions using deterministic expressions, e.g. `SELECT * FROM t WHERE length(p) > 10`. That's because, `HiveMetadata#pushdownFilter` doesn't specify any predicate when constructing Constraint object. How about I add a TODO to add logic to specify the predicate?

```
        HivePartitionResult hivePartitionResult = partitionManager.getPartitions(metastore, tableHandle, new Constraint<>(entireColumnDomain), session);
```